pkg/webhooks/customcluster_webhook_test.go

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright Kurator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhooks
+
+import (
+	"io/fs"
+	"os"
+	"path"
+	"path/filepath"
+	"testing"
+
+	. "github.com/onsi/gomega"
+	"github.com/stretchr/testify/assert"
+	"sigs.k8s.io/yaml"
+
+	"kurator.dev/kurator/pkg/apis/infra/v1alpha1"
+)
+
+func TestCustomClusterValidation(t *testing.T) {
+	// read configuration from examples directory
+	caseNames := []string{"../../examples/infra/customcluster/cc-customcluster.yaml"}
+
+	wh := &CustomClusterWebhook{}
+	for _, tt := range caseNames {
+		t.Run(tt, func(t *testing.T) {
+			g := NewWithT(t)
+			c, err := readCustomCluster(tt)
+			g.Expect(err).NotTo(HaveOccurred())
+
+			err = wh.validate(c)
+			g.Expect(err).NotTo(HaveOccurred())
+		})
+	}
+}
+
+func TestInvalidCustomClusterValidation(t *testing.T) {
+	r := path.Join("testdata", "customcluster")
+	caseNames := make([]string, 0)
+	err := filepath.WalkDir(r, func(path string, d fs.DirEntry, err error) error {
+		if d.IsDir() {
+			return nil
+		}
+
+		caseNames = append(caseNames, path)
+
+		return nil
+	})
+	assert.NoError(t, err)
+
+	wh := &CustomClusterWebhook{}
+	for _, tt := range caseNames {
+		t.Run(tt, func(t *testing.T) {
+			g := NewWithT(t)
+			c, err := readCustomCluster(tt)
+			g.Expect(err).NotTo(HaveOccurred())
+
+			err = wh.validate(c)
+			g.Expect(err).To(HaveOccurred())
```

Comment:
Also need to check which field is invalid, error message is user-oriented

Reply:
yes, the current log have this description of which field is invalid. here is the log: 

=== RUN   TestInvalidCustomClusterValidation
=== RUN   TestInvalidCustomClusterValidation/testdata\customcluster\invalid-address.yaml
    customcluster_webhook_test.go:73: CustomCluster.infrastructure.cluster.x-k8s.io "cc-customcluster" is invalid: spec.controlPlaneConfig.address: Invalid value: "192.1.1.256": invalid ip address: 192.1.1.256
=== RUN   TestInvalidCustomClusterValidation/testdata\customcluster\invalid-cni-type.yaml
    customcluster_webhook_test.go:73: CustomCluster.infrastructure.cluster.x-k8s.io "cc-customcluster" is invalid: spec.cni.type: Invalid value: "ciilium": invalid CNI type: ciilium ,it should be one of [calico canal cilium flannel kube-ovn kube-router macvlan weave] 
=== RUN   TestInvalidCustomClusterValidation/testdata\customcluster\invalid-machineref-name.yaml
    customcluster_webhook_test.go:73: CustomCluster.infrastructure.cluster.x-k8s.io "cc-customcluster" is invalid: spec.machineRef.name: Invalid value: "cc_custommachine": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
=== RUN   TestInvalidCustomClusterValidation/testdata\customcluster\invalid-machineref-namespace.yaml
    customcluster_webhook_test.go:73: CustomCluster.infrastructure.cluster.x-k8s.io "cc-customcluster" is invalid: spec.machineRef.namespace: Invalid value: "defaultdefaultdefaultdefaultdefaultdefaultdefaultdefaultdefaultt": must be no more than 63 characters
=== RUN   TestInvalidCustomClusterValidation/testdata\customcluster\miss-cni-type.yaml
    customcluster_webhook_test.go:73: CustomCluster.infrastructure.cluster.x-k8s.io "cc-customcluster" is invalid: spec.cni.type: Invalid value: "": invalid CNI type:  ,it should be one of [calico canal cilium flannel kube-ovn kube-router macvlan weave] 
=== RUN   TestInvalidCustomClusterValidation/testdata\customcluster\miss-machineref-all-value.yaml
    customcluster_webhook_test.go:73: CustomCluster.infrastructure.cluster.x-k8s.io "cc-customcluster" is invalid: [spec.machineRef.kind: Required value: must be set, spec.machineRef.apiVersion: Required value: must be set, spec.machineRef.name: Required value: must be set, spec.machineRef.namespace: Required value: must be set]
=== RUN   TestInvalidCustomClusterValidation/testdata\customcluster\miss-machineref-name.yaml
    customcluster_webhook_test.go:73: CustomCluster.infrastructure.cluster.x-k8s.io "cc-customcluster" is invalid: spec.machineRef.name: Required value: must be set
--- PASS: TestInvalidCustomClusterValidation (0.01s)